File: services/server/src/protocol/protocol.py
import socket
import safe_socket

import lottery

CMD_CLIENT_AGENCY_ID = 0x01
CMD_CLIENT_BET = 0x02
CMD_SERVER_ACK = 0x03
CMD_CLIENT_FINISHED = 0x04
CMD_CLIENT_RESULTS = 0x05
CMD_SERVER_RESULTS = 0x06

class Protocol:
    def __init__(self):
        pass

    def serializeByte(self, value: int) -> bytes:
        return value.to_bytes(1)

    def serialize2Bytes(self, value: int) -> bytes:
        return value.to_bytes(2, byteorder="big")

    def serialize4Bytes(self, value: int) -> bytes:
        return value.to_bytes(4, byteorder="big")

    def serializeString(self, value: str) -> bytes:
        encoded = value.encode("utf-8")
        length = len(encoded)
        return self.serialize2Bytes(length) + encoded

    def serializeBet(self, bet: lottery.Bet) -> bytes:
        data = b""
        data += self.serializeString(bet.first_name)
        data += self.serializeString(bet.last_name)
        data += self.serialize4Bytes(bet.document)
        data += self.serializeString(bet.birthdate)
        data += self.serialize4Bytes(bet.number)
        return data

    def readByte(self, socket: socket.socket) -> int:
        data = safe_socket.recv_all(socket, 1)
        return int.from_bytes(data)

    def read2Bytes(self, socket: socket.socket) -> int:
        data = safe_socket.recv_all(socket, 2)
        return int.from_bytes(data, byteorder="big")

    def read4Bytes(self, socket: socket.socket) -> int:
        data = safe_socket.recv_all(socket, 4)
        return int.from_bytes(data, byteorder="big")

    def readString(self, socket: socket.socket) -> str:
        length = self.read2Bytes(socket)
        data = safe_socket.recv_all(socket, length)
        return data.decode("utf-8")

    def readBet(self, socket: socket.socket, agency_id: int) -> lottery.Bet:
        first_name = self.readString(socket)
        last_name = self.readString(socket)
        document = self.read4Bytes(socket)
        birth_date = self.readString(socket)
        number = self.read4Bytes(socket)
        return lottery.Bet(agency_id, first_name, last_name, document, birth_date, number)

    def readBets(self, socket: socket.socket, agency_id: int) -> list[lottery.Bet]:
        num_bets = self.read2Bytes(socket)
        bets = []
        for _ in range(num_bets):
            bet = self.readBet(socket, agency_id)
            bets.append(bet)
        return bets

    def sendAck(self, socket: socket.socket) -> None:
        data = self.serializeByte(CMD_SERVER_ACK)
        safe_socket.send_all(socket, data)

    def sendResults(self, socket: socket.socket, results: list[lottery.Bet]) -> None:
        data = self.serializeByte(CMD_SERVER_RESULTS)
        num_bets = len(results)
        data += self.serialize2Bytes(num_bets)
        socket.sendall(data)
    
        for bet in results:
            bet_winning_data = self.serializeBet(bet)
            socket.sendall(bet_winning_data)

File: services/client/src/protocol/protocol.go
package protocol

import (
	"encoding/binary"
	"io"
	"strconv"

	"github.com/7574-sistemas-distribuidos/tp-nivelador/src/bet"
	"github.com/7574-sistemas-distribuidos/tp-nivelador/src/safe_socket"
)

const (
	CMD_CLIENT_AGENCY_ID uint8 = 1
	CMD_CLIENT_BET       uint8 = 2
	CMD_SERVER_ACK       uint8 = 3
	CMD_CLIENT_FINISHED  uint8 = 4
	CMD_CLIENT_RESULTS   uint8 = 5
	CMD_SERVER_RESULTS   uint8 = 6
)

type Protocol struct {
}

func (p *Protocol) serializeByte(value uint8) []byte {
	data := make([]byte, 1)
	data[0] = value

	return data
}

func (p *Protocol) serialize2Bytes(value uint16) []byte {
	data := make([]byte, 2)
	binary.BigEndian.PutUint16(data, value)

	return data
}

func (p *Protocol) serialize4Bytes(value uint32) []byte {
	data := make([]byte, 4)
	binary.BigEndian.PutUint32(data, value)

	return data
}

func (p *Protocol) serializeString(value string) []byte {
	data := []byte(value)
	length := uint16(len(data))
	return append(p.serialize2Bytes(length), data...)
}

func (p *Protocol) serializeBet(bet *bet.Bet) []byte {
	var data []byte

	data = append(data, p.serializeString(bet.FirstName)...)
	data = append(data, p.serializeString(bet.LastName)...)
	data = append(data, p.serialize4Bytes(uint32(bet.Document))...)
	data = append(data, p.serializeString(bet.BirthDate)...)
	data = append(data, p.serialize4Bytes(uint32(bet.BetNumber))...)

	return data
}

func (p *Protocol) readByte(socket io.Reader) (uint8, error) {

	data, err := safe_socket.RecvAll(socket, 1)
	if err != nil {
		return 0, err
	}
	return data[0], nil
}

func (p *Protocol) read2Bytes(socket io.Reader) (uint16, error) {
	data, err := safe_socket.RecvAll(socket, 2)
	if err != nil {
		return 0, err
	}
	return binary.BigEndian.Uint16(data), nil
}

func (p *Protocol) read4Bytes(socket io.Reader) (uint32, error) {
	data, err := safe_socket.RecvAll(socket, 4)
	if err != nil {
		return 0, err
	}
	return binary.BigEndian.Uint32(data), nil
}

func (p *Protocol) readString(socket io.Reader) (string, error) {
	length, err := p.read2Bytes(socket)
	if err != nil {
		return "", err
	}

	data, err := safe_socket.RecvAll(socket, int(length))
	if err != nil {
		return "", err
	}

	return string(data), nil
}

func (p *Protocol) readBet(socket io.Reader) (*bet.Bet, error) {
	firstName, err := p.readString(socket)
	if err != nil {
		return nil, err
	}

	lastName, err := p.readString(socket)
	if err != nil {
		return nil, err
	}

	document, err := p.read4Bytes(socket)
	if err != nil {
		return nil, err
	}

	birthDate, err := p.readString(socket)
	if err != nil {
		return nil, err
	}

	betNumber, err := p.read4Bytes(socket)
	if err != nil {
		return nil, err
	}

	bet := &bet.Bet{
		FirstName: firstName,
		LastName:  lastName,
		Document:  int(document),
		BirthDate: birthDate,
		BetNumber: int(betNumber),
	}
	return bet, nil
}

func (p *Protocol) ReadCommand(socket io.Reader) (uint8, error) {
	cmd, err := p.readByte(socket)
	if err != nil {
		return 0, err
	}
	return cmd, nil
}

func (p *Protocol) ReadResults(socket io.Reader) ([]*bet.Bet, error) {
	numBets, err := p.read2Bytes(socket)
	if err != nil {
		return nil, err
	}

	bets := make([]*bet.Bet, 0, numBets)
	for i := 0; i < int(numBets); i++ {
		bet, err := p.readBet(socket)
		if err != nil {
			return nil, err
		}
		bets = append(bets, bet)
	}

	return bets, nil
}

func (p *Protocol) SendAgencyId(socket io.Writer, id string) error {
	var data []byte

	value, err := strconv.Atoi(id)
	if err != nil {
		return err
	}

	cmd := p.serializeByte(CMD_CLIENT_AGENCY_ID)
	agencyId := p.serializeByte(uint8(value))
	data = append(data, cmd...)
	data = append(data, agencyId...)
	err = safe_socket.SendAll(socket, data)
	return err
}

func (p *Protocol) SendBets(socket io.Writer, bets *[]bet.Bet) error {
	var data []byte
	cmd := p.serializeByte(CMD_CLIENT_BET)
	numBets := p.serialize2Bytes(uint16(len(*bets)))
	data = append(data, cmd...)
	data = append(data, numBets...)
	for _, bet := range *bets {
		betData := p.serializeBet(&bet)
		data = append(data, betData...)
	}
	err := safe_socket.SendAll(socket, data)
	return err
}

func (p *Protocol) SendCommandFinished(socket io.Writer) error {
	cmd := p.serializeByte(CMD_CLIENT_FINISHED)
	err := safe_socket.SendAll(socket, cmd)
	return err
}

func (p *Protocol) SendCommandResults(socket io.Writer) error {
	cmd := p.serializeByte(CMD_CLIENT_RESULTS)
	err := safe_socket.SendAll(socket, cmd)
	return err
}
